### DISCUSSION TITLE
Ability to link a third (or even more) tables into a many-to-many relationship intersection table

### DISCUSSION THREAD

**Discussion by jwleon** [directus/directus#2700] [Feature Requests]
I quite often have the case where I need to "attribute" a connection between two tables. If I had to name it (maybe there already is a professional term for it) I would call it

**many-to-many relationship with linked attribution**

Let's say I have a collection `projects` and a the standard table `directus_users`, which I would like to link in a way that I can assign many users to many projects. That is no problem at all. I get an intersection table called `projects_directus_users`.

But let's bring into the example a third collection/table called `positions` - which contains an extendable list of roles that people can have in projects (like organizer, participant, ...). If I were to design the database myself, I would just create a _third column inside of my intersection table_ called `positions_id`. While I can do it manually in MySQL on the Directus tables, I found no way in Directus to input and display this additional information. If I didn't miss anything and it really isn't possible, I would like to suggest this feature for an upcoming release.

I guess it is more or less a question of writing an interface that can handle that - not so much a groundbreaking change in architecture of Directus... Unfortunately, after looking into extension architecture of Directus, I am too far away from understanding the programming to hack somehing together and suggest it...

**Comment by benhaynes** [comment]
Hey! We call this type of relationship a `Many-to-Any` (M2X). It's on our roadmap, and is a high-priority for the upcoming version 9 release. It is similar to a M2M, but allows you to relate items from _different_ collections. This opens the door to a lot of powerful interfaces, such as "modular page builders".

Docs: https://docs.directus.io/guides/relationships.html#many-to-any
Ticket: https://github.com/directus/api/issues/8

**Reply by hulkthedev** [reply]
New Docs link? 404

**Reply by benhaynes** [reply]
https://docs.directus.io/configuration/relationships/#many-to-any-m2a

Also, the docs are searchable... all I did was search for `many` 😉

**Comment by jwleon** [comment]
Thanks for the response! As of what I've read in the documentation, I don't think it is a many-to-any relationship, but I found other topics here referencing to it as many-to-many-many relationship. But I guess there are a lot of similarities and it looks like what you are up to for the upcoming releases will also cover my use case. Looking forward to it and thanks for your great work!

**Reply by benhaynes** [reply]
The other option is to have additional fields on the Junction Collection that are included in the parent form — this is also something we're working towards in upcoming versions.

**Reply by rijkvanzanten** [reply]
@jwleon you're right!

@benhaynes what we're talking about is saving the following in the junction table:

```
junction
-------------------
id
current_id
related_id
related_collection
```

that way, we can save any value in `related_id` and use the `related_collection` to know what collection that item points to. That allows us to do things like "page builders" where you select "blocks" in order.

What @jwleon is talking about is slightly different though, they would like to be able to use multiple related items on the same junction table:

```
junction
---------------------
id
user_id
project_id
position_id
```

This doesn't have the need to save what collection to pull from, as both `project_id` and `position_id` always point to the same collection.

**Reply by rijkvanzanten** [reply]
That being said, like you also mentioned @benhaynes, I believe adding a many to one field on the junction table, and have that show up in the edit modal in a m2m achieves this 🙂

**Reply by chishaku** [reply]
> @benhaynes what we're talking about is saving the following in the junction table:
> 
> ```
> junction
> -------------------
> id
> current_id
> related_id
> related_collection
> ```
> 
> that way, we can save any value in `related_id` and use the `related_collection` to know what collection that item points to. That allows us to do things like "page builders" where you select "blocks" in order.
> 


@rijkvanzanten 

Is a o2m relationship required on the `related_collection` to query all the related `current_id`'s?

Or is the m2a relationship on `current` supposed to be sufficient to query related objects in both directions?

**Comment by rijkvanzanten** [comment]
This is now available starting with `v9.0.0-rc.19` ✨

**Comment by sxsws** [comment]
The M2A does not solve the problem of @jwleon. Op means to have a junction table like (employee_id, department_id, role_id), where the junction table has more than one relation for each row, and not to allow the junction table to link to other model types.

This would allow an employee to be put in a role per department. and there could never be two records of the same employee, department and role id. However, the employee could have the same role in a different dept, or a different role in the same dept, or even a different employee could have the same role in the same dept.

Junction tables with multiple relations like this are an important thing.

**Reply by rijkvanzanten** [reply]
You can add any additional fields to the junction table to your heart's content. The piece to set a unique constraint across multiple fields is discussed under [the composite key's discussion](https://github.com/directus/directus/discussions?discussions_q=composite)
